tiledb/sm/rest/rest_client.cc

```diff
@@ -528,6 +528,155 @@ Status RestClient::submit_query_to_rest(const URI& uri, Query* query) {
+    uint64_t buffer_cells = *query_buffer.buffer_size_ / cache.cell_size;
+    // Only consider cached data up to byte_offset that is queued for submit.
+    uint64_t cache_buffer_cells = cache.buffer.size() / cache.cell_size;
+    uint64_t total_buffer_cells = buffer_cells + cache_buffer_cells;
+
+    // If buffer_cells does not fill a tile, cache the entire buffer.
+    uint64_t tile_cache_cells = buffer_cells;
+    uint64_t buffer_cache_bytes = tile_cache_cells * cache.cell_size;
+    // Check if data between cache and user buffer is tile-aligned.
+    if (total_buffer_cells >= cell_num_per_tile &&
+        total_buffer_cells % cell_num_per_tile == 0) {
+      // Submit all data and cache nothing after submit.
+      cache.cache_bytes = 0;
+      continue;
+    }
+
+    // Cases that will not write data into the cache.
+    bool cache_data = true;
+    if (buffer_cells >= cell_num_per_tile) {
+      tile_cache_cells =
+          (cache_buffer_cells + buffer_cells) % cell_num_per_tile;
+      buffer_cache_bytes = tile_cache_cells * cache.cell_size;
+      cache_data = false;
+    } else if (total_buffer_cells >= cell_num_per_tile) {
+      // The user buffer and cache data combined can fill one or more tiles.
+      tile_cache_cells = cell_num_per_tile % cache_buffer_cells;
+      buffer_cache_bytes = (buffer_cells - tile_cache_cells) * cache.cell_size;
+      cache_data = false;
+    }
+    uint64_t tile_cache_bytes = tile_cache_cells * cache.cell_size;
+
+    if (cache_data) {
+      // Write fixed attribute trimmed data into cache buffer.
+      throw_if_not_ok(cache.buffer.write(
+          (char*)query_buffer.buffer_ + *query_buffer.buffer_size_ -
+              buffer_cache_bytes,
+          tile_cache_bytes));
+      *query_buffer.buffer_size_ -= buffer_cache_bytes;
+
+      if (is_var) {
+        uint64_t var_buffer_size = *query_buffer.buffer_var_size_;
+        // Total bytes to cache from var data and offset buffers.
+        uint64_t var_data_cache_bytes = 0;
+        uint64_t first_cached_offset = 0;
+        if (tile_cache_cells != 0) {
+          // Use var_buffer_size if we are caching the entire var-size buffer.
+          first_cached_offset =
+              buffer_cells == tile_cache_cells ?
+                  var_buffer_size :
+                  ((uint64_t*)
+                       query_buffer.buffer_)[buffer_cells - tile_cache_cells];
+        }
+
+        // Use the first cached offset to calculate var bytes to cache.
+        var_data_cache_bytes = first_cached_offset == var_buffer_size ?
+                                   first_cached_offset :
+                                   var_buffer_size - first_cached_offset;
+
+        // Adjust any appended offsets to be relative to already cached data.
+        for (uint64_t pos = cache.buffer.size() - tile_cache_bytes;
+             pos < cache.buffer.size();
+             pos += constants::cell_var_offset_size) {
+          *cache.get_cell_offset(pos / cache.cell_size) +=
```

Comment:
get_cell_offset should go away. We should make a function that fixes offsets and includes this whole for loop.